Kernel/KBO.cpp

```diff
@@ -182,106 +181,206 @@ void KBO::State::recordVariable(unsigned var, int coef)
   }
 }
 
-void KBO::State::traverse(TermList tl,int coef)
+template<int coef, bool varsOnly>
+void KBO::State::traverse(AppliedTerm tt)
 {
-  if(tl.isOrdinaryVar()) {
-    _weightDiff += _kbo._funcWeights._specialWeights._variableWeight * coef;
-    recordVariable(tl.var(), coef);
+  static_assert(coef==1 || coef==-1);
+
+  if (tt.term.isVar()) {
+    if constexpr (!varsOnly) {
+      _weightDiff += _kbo._funcWeights._specialWeights._variableWeight * coef;
+    }
+    recordVariable<coef>(tt.term.var());
     return;
   }
-  ASS(tl.isTerm());
-
-  Term* t=tl.term();
-  ASSERT_VALID(*t);
-
-  _weightDiff+=_kbo.symbolWeight(t)*coef;
+  struct State {
+    AppliedTerm t;
+    unsigned arg;
+  };
+  static Stack<State> recState;
+  recState.push(State{ tt, 0 });
 
-  if(!t->arity()) {
-    return;
+  if constexpr (!varsOnly) {
+    _weightDiff += _kbo.symbolWeight(tt.term.term()) * coef;
   }
 
-  TermList* ts=t->args();
-  static Stack<TermList*> stack(4);
-  for(;;) {
-    if(!ts->next()->isEmpty()) {
-      stack.push(ts->next());
+  while (recState.isNonEmpty()) {
+    auto& curr = recState.top();
+    if (curr.arg >= curr.t.term.term()->arity()) {
+      recState.pop();
+      continue;
     }
-    if(ts->isTerm()) {
-      auto term = ts->term();
-      _weightDiff+=_kbo.symbolWeight(term)*coef;
-      if(term->arity()) {
-	stack.push(term->args());
+    AppliedTerm t(*curr.t.term.term()->nthArgument(curr.arg++), curr.t);
+    if (t.term.isVar()) {
+      ASS(!t.aboveVar);
+      if constexpr (!varsOnly) {
+        _weightDiff += _kbo._funcWeights._specialWeights._variableWeight * coef;
       }
-    } else {
-      ASS_METHOD(*ts,isOrdinaryVar());
-      auto var = ts->var();
-      _weightDiff += _kbo._funcWeights._specialWeights._variableWeight * coef;
-      recordVariable(var, coef);
+      recordVariable<coef>(t.term.var());
+      continue;
     }
-    if(stack.isEmpty()) {
-      break;
+
+    if constexpr (varsOnly) {
+      if (!t.term.term()->ground()) {
+        recState.push(State{ t, 0 });
+      }
+    } else {
+      _weightDiff += _kbo.symbolWeight(t.term.term()) * coef;
+      recState.push(State{ t, 0 });
     }
-    ts=stack.pop();
   }
 }
 
-void KBO::State::traverse(Term* t1, Term* t2)
+template<bool unidirectional>
+Ordering::Result KBO::State::traverseLex(AppliedTerm tl1, AppliedTerm tl2)
```

Comment:
I don't quite like the look of this function. In every part of the function it branches on `unidirectional` and there is huge chunks of code in every conditional, and the behaviour shared between the two cases `undirectional = true` and `unidirectional = false` do not have a lot in common. This makes the function even longer than it was before (and in my view it was already too long before this refactor), and thus even harder to understand what was going on. I'd suggest to have two different functions 
`traverseLexBidir` and `traverseLexUnidir` or so. 
If you really want to make use of code sharing between the two cases I'd suggest factoring out functions with appropriate names that describe what's going on in the case splits.